File: python/object-enviar.py
#!/usr/bin/env python3

import os
import sys
from argparse import ArgumentParser
import json

import boto3
from botocore.config import Config
from urllib3.exceptions import InsecureRequestWarning    
from urllib3 import disable_warnings    
disable_warnings(InsecureRequestWarning)

with open('credentials.json', 'r') as fd:
    credentials = json.loads(fd.read())

def main():
    parser = ArgumentParser(description='Enviar um objeto para um bucket')
    parser.add_argument('--bucket-name',
                        dest='bucket_name',
                        action='store',
                        required=True,
                        help='Nome do bucket que armazenará o objeto')
    parser.add_argument('--key-name',
                        dest='key_name',
                        action='store',
                        required=True,
                        help='Digite a key do objeto')
    parser.add_argument('--path',
                        dest='path',
                        action='store',
                        help='O caminho do arquivo a ser armazenado')
    parser.add_argument('--data',
                        dest='data',
                        action='store',
                        help='O dado a ser armazenado')

    args = parser.parse_args()

    if not args.path and not args.data:
        parser.error('É necessário fornecer --path ou --data')

    s3_config = Config(signature_version='s3v4')

    s3 = boto3.resource('s3',
                        endpoint_url=credentials['endpoint_url'],
                        aws_access_key_id=credentials['access_key'],
                        aws_secret_access_key=credentials['secret_key'],
                        config=s3_config,
                        use_ssl=True,
                        verify=False)

    bucket = s3.Bucket(args.bucket_name)

    if args.path:
        with open(args.path, 'rb') as file_data:
            bucket.put_object(Bucket=args.bucket_name,
                              Key=args.key_name,
                              Body=file_data)
    else:
        bucket.put_object(Bucket=args.bucket_name,
                          Key=args.key_name,
                          Body=args.data)

if __name__ == '__main__':
    main()

File: node/object-enviar.js
#!/usr/bin/env node

const { S3Client, PutObjectCommand } = require("@aws-sdk/client-s3");
const { NodeHttpHandler } = require("@aws-sdk/node-http-handler");
const fs = require('fs');
const { Command } = require('commander');
const program = new Command();

const credentials = JSON.parse(fs.readFileSync('credentials.json'));

const s3Client = new S3Client({
    endpoint: credentials.endpoint_url,
    region: "us-east-1",
    credentials: {
        accessKeyId: credentials.access_key,
        secretAccessKey: credentials.secret_key
    },
    requestHandler: new NodeHttpHandler({
        httpsAgent: new require('https').Agent({
            rejectUnauthorized: false
        })
    }),
    forcePathStyle: true  // Força o estilo de caminho para o endpoint
});

program
    .requiredOption('--bucket-name <bucketName>', 'Nome do bucket que armazenará o objeto')
    .requiredOption('--key-name <keyName>', 'Digite a key do objeto')
    .option('--path <filePath>', 'O caminho do arquivo a ser armazenado')
    .option('--data <data>', 'O dado a ser armazenado');

program.parse(process.argv);

const options = program.opts();

if (!options.path && !options.data) {
    console.error('É necessário fornecer --path ou --data');
    process.exit(1);
}

const uploadObject = async () => {
    try {
        const uploadParams = { Bucket: options.bucketName, Key: options.keyName };
        if (options.path) {
            uploadParams.Body = fs.createReadStream(options.path);
        } else {
            uploadParams.Body = options.data;
        }
        await s3Client.send(new PutObjectCommand(uploadParams));
        console.log(`Objeto ${options.keyName} enviado com sucesso para o bucket ${options.bucketName}.`);
    } catch (err) {
        console.error('Erro ao enviar o objeto:', err);
    }
};

uploadObject();

